Created basic project stub

matejak opened the pull request:
This PR creates a basic project setup featuring
* docker support for the server,
* docker stub for the database,
* docker-compose wrapper,
* basic http listener that writes stuff to a sqlite DB,
* setup script.

This PR has rough edges at time of the submission, but the good thing is that one is able to get the server running just using the `docker-compose up` command. The server is able to process GET requests s.a. `http://localhost:8080/?name=me&msg=something`

matejak committed:
Added basic project skeleton

matejak committed:
Added basic stub of database and GET requests processing

matejak committed:
Fixed the setup.py script

petrjoachim commented:
Could you please review your commit messages after reading through this guideline:
http://api.coala.io/en/latest/Developers/Writing_Good_Commits.html

Also I've created #13 in order to be aligned in the style of commits and other things.

petrjoachim commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,12 @@
+    ports:
+     - "8080:8080"
+  database:
+    build: ./database
```
It's not necessary to have separate Dockerfile for postgres as it doesn't add anything to it.

matejak replied:
The postgres container is not used so far. Judging from the image description (https://hub.docker.com/_/postgres/), I thought that we will end up with our Dockerfile extending it. I suppose that we will clarify this on our upcoming mentoring session.

petrjoachim commented on docker/server/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+	&& ln -s /iwant-bot-cloned /iwant-bot \
+	&& (cd iwant-bot && git checkout docker-support) \
+	&& pip install -r iwant-bot/requirements.txt \
+	&& true
```
This part should not clone from remote origin.
Instead it should `ADD` files from local storage and then run installation on top of them

matejak replied:
I am now thinking of two Docker images - one for reproducible builds (i.e. it would download releases from GitHub) and one for development (here I find a shared volume more flexible than copying a directory with source files). Let's discuss this in the upcoming mentoring session.

petrjoachim commented on docker/server/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+	&& git clone https://github.com/kiwicom/iwant-bot.git /iwant-bot-cloned \
+	&& ln -s /iwant-bot-cloned /iwant-bot \
+	&& (cd iwant-bot && git checkout docker-support) \
+	&& pip install -r iwant-bot/requirements.txt \
```
We should use `setup.py` when we have it

matejak replied:
This sounds legit.

petrjoachim commented on docker/server/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+	&& pip install -r iwant-bot/requirements.txt \
+	&& true
+
+CMD [ "python", "/iwant-bot/src/iwant_bot/start.py"]
```
Also when we have python package we should start it like this`python -m iwant_bot`

matejak committed:
Made the project installable using setup.py

The package now can be installed using setup.py install/develop
in such way that it can be launched using `python -m iwant_bot.start`
The server Dockerfile has been updated to reflect this.

iakat commented on docker/server/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM python:3.6
```
Use `python:3.6-alpine` or if not possible, `python:3.6-slim`

iakat replied:
also, move to the root directory, `/Dockerfile`

matejak replied:
done, `alpine` is feasible if we don't use `apt-get`, which we don't really need.

iakat commented on docker/server/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM python:3.6
+
+RUN true \
```
can we get rid of this? only makes the file longer

matejak replied:
I like this construct, because it is friendly both to SCMs and to some text editors s.a. vim. If it disturbs you, I can of course remove it.

iakat commented on docker/server/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+	&& apt-get install -y --no-install-recommends \
+		git vim-tiny \
+	&& apt-get clean \
+	&& true
```
replace this with `rm -rf /var/lib/apt/lists/* /tmp/*`

iakat commented on docker/server/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+	&& apt-get clean \
+	&& true
+
+RUN true \
```
same as above

iakat commented on docker/server/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+	&& true
+
+RUN true \
+	&& git clone https://github.com/kiwicom/iwant-bot.git /iwant-bot-cloned \
```
don't we want to have `COPY` to get files from the repo to the build stage? something like https://github.com/kiwicom/starpicker/blob/master/Dockerfile#L5-L10

iakat commented on docker/server/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+	&& (cd iwant-bot && python setup.py develop) \
+	&& true
+
+CMD [ "python", "python -m iwant_bot.start"]
```
looks wrong

matejak replied:
Indeed. It has been fixed and it uses guicorn with the auto-reload feature, as the mentor has suggested.

iakat commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,12 @@
+version: '3'
```
dubs for using version 3! but can we move the file in the root of the repo? /docker-compose.yml

iakat commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,12 @@
+    build: ./server
+    image: "iwantbot/server"
+    volumes:
+     - ..:/iwant-bot-mount
```
replace with `/app` as suggested below

matejak replied:
done.

iakat commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,12 @@
+services:
+  server:
+    build: ./server
+    image: "iwantbot/server"
```
`kiwicom/iwant-bot`, we have a docker hub user: https://hub.docker.com/r/kiwicom/

matejak committed:
Normalized Docker support files.

Docekrfiles and docker-compose.yml have been moved to project root.
Postgres Dockerfile was dropped.

matejak committed:
requirements.txt introduced properly

requirements.txt is generated from requirements.in (newly committed to the repo).

matejak committed:
Modified the project so it is gunicorn-compatible.

Also updated README.md accordingly.

matejak commented on docker/server/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM python:3.6
+
+RUN true \
```
done.

iakat commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+
+COPY requirements.txt /app/
+
+RUN true \
```
still don't like this

matejak replied:
OK, so be it, I will remove it in the next commit.

iakat commented on docker-compose.yml:
```diff
@@ -0,0 +1,11 @@
+services:
+  server:
+    build: .
+    image: "iwantbot/server"
```
kiwicom/iwantbot

matejak replied:
My bad, I have thought that I have fixed this. Will do in the next commit.

iakat commented on setup.py:
```diff
@@ -0,0 +1,99 @@
+    # Versions should comply with PEP440.  For a discussion on single-sourcing
+    # the version across setup.py and the project code, see
+    # https://packaging.python.org/en/latest/single_source_version.html
+    version='0.0.0',
```
0.1.0

matejak replied:
I don't know how we would like to manage the version number in a smart way - quite honestly, I would like to make a decision after consulting mentors.

iakat replied:
semver sounds good here

matejak replied:
I am all in for semver, but I meant this aspect of versioning: https://packaging.python.org/single_source_version/

chauffer commented:
please get acquainted with https://coala.io/commit - probably good enough to have this all in one commit for now, though

iakat commented on .gitignore:
```diff
@@ -87,3 +87,6 @@ ENV/
 # Rope project settings
 .ropeproject
+
+# File backups
```
out of scope for the initial commit adding the whole project

petrjoachim replied:
let's ignore this for now

iakat commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+COPY requirements.txt /app/
+
+RUN true \
+	&& pip install gunicorn \
```
gunicorn should be in requirements.txt

matejak replied:
I see it as a dependency of the image rather of the package.
It is not imported anywhere in the Python code. To run the app using gunicorn is a decision on the image level, it could be accomplished otherwise (i.e. using just the python interpreter).

iakat commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+	&& pip install -r /app/requirements.txt \
+	&& true
+
+WORKDIR /app/src
```
why `/app/src` and not `/app`? 🤔

matejak replied:
See the comment above. The `iwant_bot` package is located in the `src` subdirectory.

iakat replied:
keep `/app` please, we're not going to store anything other than the source

matejak replied:
The decision of the project folder structure should be made on a mentoring session (so others can learn why it is good to do it in a particular way). If we decide there that it is better to have another directory structure, I (or somebody else) will move it accordingly to the decision.

iakat replied:
ok, this is popular docker conventions.

iakat replied:
ok, this is popular docker conventions.

iakat commented on README.md:
```diff
@@ -3,4 +3,11 @@ Slackbot project implemented by [Kiwi.com](Kiwi.com) interns
+
+## How to set up
+
+* Make sure you have `pip-tools`, `docker` and `docker-compose`.
```
`pip-tools` doesn't seem needed for running

matejak replied:
The next lines explain why it is suggested - the image building uses requirements.txt, which can be obtained using `pip-compile`. In theory, the requirements.txt file should be in a good shape, but there is a low, but non-zero probability that a person that wants to build the image may need to regenerate it (e.g. when a particular version of a dependency is deleted from pypi).
I agree that it is a dependency for building with low likelihood of being needed, so if the general consensus is to remove this mention, I am comfortable removing it.

iakat replied:
can we list edge cases outside of the "How to set up" block, then?

matejak replied:
OK.

iakat commented on docker-compose.yml:
```diff
@@ -0,0 +1,11 @@
+    volumes:
+     - .:/app
+    ports:
+     - "8080:8080"
```
make it listen on port `80` in the container, no need to not follow the standards when we're in a container; we have 80 available all the time

matejak replied:
I agree.

iakat commented on src/iwant_bot/start.py:
```diff
@@ -0,0 +1,32 @@
+
+    messages_we_got_so_far = DB_ACCESS.load_n_last_messages()
+    text = ["We recognize the 'name' and 'msg' GET query attributes,"]
+    text += ["so you can do http://localhost:8080/?name=me&msg=message"]
```
use `'` everywhere, except when you'd have to escape `'`: use `"` only then

matejak replied:
OK, I will do that.

iakat commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+	&& pip install gunicorn \
+	&& pip install -r /app/requirements.txt \
+	&& true
+
```
where are we copying the app and installing it?

matejak replied:
We are not installing, the app directory is mounted to `/app`. Doing it this way removes necessity of installation and using `gunicorn` ensures that when the dev modifies app file on the host, `gunicorn` will reload the app.

iakat replied:
it can and very much should be copied, otherwise this Dockerfile is useless.
the mounst will overwrite the content inside the image

matejak replied:
I respectfuly disagree. This form of Dockerfile is a result of the mentoring session with Jiří. The container is intended to be used by developers to be able to quickly test their code, and I can confirm that it works as intended. As such, it is not useless.

iakat replied:
there is no reason to not have the app in the container. the mount has priority over it.

iakat commented on requirements.in:
```diff
@@ -0,0 +1,2 @@
+aiohttp
+sqlalchemy
```
`psycopg2` needed as well for the postgres transport

matejak replied:
It is not needed at the moment as we don't use postgres yet.
However, thank you for heads-up, we plan to use postgres as soon as the other basics get sorted out.

iakat replied:
in that case, get rid of postgres in `docker-compose.yml` for now, or implement postgres (preferred)

matejak replied:
OK. This was meant to be a fast PR just to get something to start with, so I will reduce the docker-compose.yml in order to get going.

matejak committed:
Removed the database reference

The app now uses sqlite, so referencing the postgres image is redundant.

matejak committed:
Changed the container port the app listens at

It went from 8080 -> 80. The host port is still at 8080 though, so the
end-user experience is the same.

matejak committed:
Restructured README.md regarding corner cases

Corner cases are no longer part of the general setup instructions.

matejak committed:
Updated style of the RUN command in the Dockerfile

matejak committed:
Changed quotes from " to ' in the Python app.

matejak committed:
Changed the iwantbot-server app image name

petrjoachim commented on Dockerfile:
```diff
@@ -0,0 +1,10 @@
+
+RUN pip install gunicorn \
+	&& pip install -r /app/requirements.txt
+
```
COPY whole project 😞

matejak committed:
address PR requests
